Auto-generate config struct from parameters.yaml and provide REST API

haoming29 opened the pull request:
Per #225 , This PR introduce a new origin RESTful API `/api/v1.0/config` that returns the configurations of the server. It reads the `docs/parameter.yaml` and parse it as a Go struct in `param/parameter_struct.go`. Later at the server/client init stage (InitServer/InitClient), it unmarshal Viper config into such struct and made available to UI via RESTful API.

A note to the reviewer, right now I placed the API at `origin_ui/origin.go` but we might want to put this route to `web_ui/ui.go` in the future given that this is a generic API for all servers. The reason I did this is that there seems to be no authentication middleware in `web_ui` and I don't want to expose server config to unauthorized user. In the long term, we should figure out auth mechanism for `web_ui`

CannonLock commented:
@haoming29 I got the following when I ran the website on your branch:

```{
  "ConfigBase": "",
  "Debug": false,
  "DisableHttpProxy": false,
  "DisableProxyFallback": false,
  "IssuerKey": "/etc/pelican/issuer.jwk",
  "MinimumDownloadSpeed": 0,
  "TLSSkipVerify": true,
  "Client": {
    "DisableHttpProxy": false,
    "DisableProxyFallback": false,
    "MinimumDownloadSpeed": 0,
    "StoppedTransferTimeout": 0,
    "SlowTransferRampupTime": 0,
    "SlowTransferWindow": 0
  },
  "Director": {
    "GeoIPLocation": "/var/cache/pelican/maxmind/GeoLite2-City.mmdb",
    "DefaultResponse": "cache",
    "MaxMindKeyFile": ""
  },
  "Federation": {
    "NamespaceUrl": "",
    "JwkUrl": "",
    "DiscoveryUrl": "https://osg-htc.org",
    "TopologyNamespaceUrl": "https://topology.opensciencegrid.org/stashcache/namespaces.json",
    "DirectorUrl": ""
  },
  "Monitoring": {
    "DataLocation": "/var/lib/pelican/monitoring/data",
    "PortLower": 9930,
    "PortHigher": 9999
  },
  "OIDC": {
    "ClientIDFile": "/etc/pelican/oidc-client-id",
    "ClientSecretFile": "/etc/pelican/oidc-client-secret",
    "DeviceAuthEndpoint": "",
    "TokenEndpoint": "",
    "UserInfoEndpoint": ""
  },
  "Origin": {
    "NamespacePrefix": "",
    "Multiuser": false,
    "UseCmsd": false,
    "UIPasswordFile": "/etc/pelican/origin-ui-passwd",
    "SelfTest": true,
    "Url": "https://d92baeb5b3f7:8443",
    "ExportVolume": "/tmp/stash/:/test"
  },
  "Registry": {
    "DbLocation": "/var/lib/pelican/registry.sqlite"
  },
  "Server": {
    "TLSKey": "/etc/pelican/certificates/tls.key",
    "Port": 8444,
    "Address": "0.0.0.0",
    "IssuerJwks": "",
    "TLSCertificate": "/etc/pelican/certificates/tls.crt",
    "TLSCACertificateFile": "/etc/pki/tls/cert.pem",
    "TLSCACertificateDirectory": "",
    "TLSCAKey": "/etc/pelican/certificates/tlsca.key",
    "ExternalAddress": "",
    "Hostname": "d92baeb5b3f7"
  },
  "Transport": {
    "ExpectContinueTimeout": 1000000000,
    "ResponseHeaderTimeout": 10000000000,
    "DialerTimeout": 10000000000,
    "DialerKeepAlive": 30000000000,
    "MaxIdleConns": 30,
    "IdleConnTimeout": 90000000000,
    "TLSHandshakeTimeout": 15000000000
  },
  "Xrootd": {
    "RobotsTxtFile": "/etc/pelican/robots.txt",
    "MacaroonsKeyFile": "/etc/pelican/macaroons-secret",
    "Authfile": "/etc/pelican/xrootd/authfile",
    "ManagerHost": "redirector.osgstorage.org",
    "SummaryMonitoringHost": "xrd-report.osgstorage.org",
    "DetailedMonitoringHost": "xrd-mon.osgstorage.org",
    "Port": 8443,
    "RunLocation": "/run/pelican/xrootd",
    "ScitokensConfig": "/etc/pelican/xrootd/scitokens.cfg",
    "Mount": "",
    "LocalMonitoringHost": "",
    "Sitename": "d92baeb5b3f7"
  }
}```

CannonLock commented:
Also if it is not already this api should be behind the security system that Emma set up.

haoming29 commented on origin_ui/origin.go:
```diff
@@ -389,6 +399,10 @@ func ConfigureOriginUI(router *gin.Engine) error {
 
+	// Register using router now. Can be converted to a group when more routes
+	// are added
+	router.GET("/api/v1.0/config", authHandler, getConfig)
```
@CannonLock The API is behind the auth

haoming29 commented:
> @haoming29 I got the following when I ran the website on your branch. Should ConfigBase have a default value?
> 
> ```
>   "ConfigBase": "",
>   "Debug": false,
>   "DisableHttpProxy": false,
>   "DisableProxyFallback": false,
>   "IssuerKey": "/etc/pelican/issuer.jwk",
>   "MinimumDownloadSpeed": 0,
>   "TLSSkipVerify": true,
>   "Client": {
>     "DisableHttpProxy": false,
>     "DisableProxyFallback": false,
>     "MinimumDownloadSpeed": 0,
>     "StoppedTransferTimeout": 0,
>     "SlowTransferRampupTime": 0,
>     "SlowTransferWindow": 0
>   },
>   "Director": {
>     "GeoIPLocation": "/var/cache/pelican/maxmind/GeoLite2-City.mmdb",
>     "DefaultResponse": "cache",
>     "MaxMindKeyFile": ""
>   },
>   "Federation": {
>     "NamespaceUrl": "",
>     "JwkUrl": "",
>     "DiscoveryUrl": "https://osg-htc.org",
>     "TopologyNamespaceUrl": "https://topology.opensciencegrid.org/stashcache/namespaces.json",
>     "DirectorUrl": ""
>   },
>   "Monitoring": {
>     "DataLocation": "/var/lib/pelican/monitoring/data",
>     "PortLower": 9930,
>     "PortHigher": 9999
>   },
>   "OIDC": {
>     "ClientIDFile": "/etc/pelican/oidc-client-id",
>     "ClientSecretFile": "/etc/pelican/oidc-client-secret",
>     "DeviceAuthEndpoint": "",
>     "TokenEndpoint": "",
>     "UserInfoEndpoint": ""
>   },
>   "Origin": {
>     "NamespacePrefix": "",
>     "Multiuser": false,
>     "UseCmsd": false,
>     "UIPasswordFile": "/etc/pelican/origin-ui-passwd",
>     "SelfTest": true,
>     "Url": "https://d92baeb5b3f7:8443",
>     "ExportVolume": "/tmp/stash/:/test"
>   },
>   "Registry": {
>     "DbLocation": "/var/lib/pelican/registry.sqlite"
>   },
>   "Server": {
>     "TLSKey": "/etc/pelican/certificates/tls.key",
>     "Port": 8444,
>     "Address": "0.0.0.0",
>     "IssuerJwks": "",
>     "TLSCertificate": "/etc/pelican/certificates/tls.crt",
>     "TLSCACertificateFile": "/etc/pki/tls/cert.pem",
>     "TLSCACertificateDirectory": "",
>     "TLSCAKey": "/etc/pelican/certificates/tlsca.key",
>     "ExternalAddress": "",
>     "Hostname": "d92baeb5b3f7"
>   },
>   "Transport": {
>     "ExpectContinueTimeout": 1000000000,
>     "ResponseHeaderTimeout": 10000000000,
>     "DialerTimeout": 10000000000,
>     "DialerKeepAlive": 30000000000,
>     "MaxIdleConns": 30,
>     "IdleConnTimeout": 90000000000,
>     "TLSHandshakeTimeout": 15000000000
>   },
>   "Xrootd": {
>     "RobotsTxtFile": "/etc/pelican/robots.txt",
>     "MacaroonsKeyFile": "/etc/pelican/macaroons-secret",
>     "Authfile": "/etc/pelican/xrootd/authfile",
>     "ManagerHost": "redirector.osgstorage.org",
>     "SummaryMonitoringHost": "xrd-report.osgstorage.org",
>     "DetailedMonitoringHost": "xrd-mon.osgstorage.org",
>     "Port": 8443,
>     "RunLocation": "/run/pelican/xrootd",
>     "ScitokensConfig": "/etc/pelican/xrootd/scitokens.cfg",
>     "Mount": "",
>     "LocalMonitoringHost": "",
>     "Sitename": "d92baeb5b3f7"
>   }
> }```
> ```

Surprisingly, it seems that we never set the value of ConfigBase, even the default value in our code. Will ask other people why this is happening but it should be a bug not related to this PR. Thanks for catching it though!

turetske commented:
> > @haoming29 I got the following when I ran the website on your branch. Should ConfigBase have a default value?
> > ```
> >   "ConfigBase": "",
> >   "Debug": false,
> >   "DisableHttpProxy": false,
> >   "DisableProxyFallback": false,
> >   "IssuerKey": "/etc/pelican/issuer.jwk",
> >   "MinimumDownloadSpeed": 0,
> >   "TLSSkipVerify": true,
> >   "Client": {
> >     "DisableHttpProxy": false,
> >     "DisableProxyFallback": false,
> >     "MinimumDownloadSpeed": 0,
> >     "StoppedTransferTimeout": 0,
> >     "SlowTransferRampupTime": 0,
> >     "SlowTransferWindow": 0
> >   },
> >   "Director": {
> >     "GeoIPLocation": "/var/cache/pelican/maxmind/GeoLite2-City.mmdb",
> >     "DefaultResponse": "cache",
> >     "MaxMindKeyFile": ""
> >   },
> >   "Federation": {
> >     "NamespaceUrl": "",
> >     "JwkUrl": "",
> >     "DiscoveryUrl": "https://osg-htc.org",
> >     "TopologyNamespaceUrl": "https://topology.opensciencegrid.org/stashcache/namespaces.json",
> >     "DirectorUrl": ""
> >   },
> >   "Monitoring": {
> >     "DataLocation": "/var/lib/pelican/monitoring/data",
> >     "PortLower": 9930,
> >     "PortHigher": 9999
> >   },
> >   "OIDC": {
> >     "ClientIDFile": "/etc/pelican/oidc-client-id",
> >     "ClientSecretFile": "/etc/pelican/oidc-client-secret",
> >     "DeviceAuthEndpoint": "",
> >     "TokenEndpoint": "",
> >     "UserInfoEndpoint": ""
> >   },
> >   "Origin": {
> >     "NamespacePrefix": "",
> >     "Multiuser": false,
> >     "UseCmsd": false,
> >     "UIPasswordFile": "/etc/pelican/origin-ui-passwd",
> >     "SelfTest": true,
> >     "Url": "https://d92baeb5b3f7:8443",
> >     "ExportVolume": "/tmp/stash/:/test"
> >   },
> >   "Registry": {
> >     "DbLocation": "/var/lib/pelican/registry.sqlite"
> >   },
> >   "Server": {
> >     "TLSKey": "/etc/pelican/certificates/tls.key",
> >     "Port": 8444,
> >     "Address": "0.0.0.0",
> >     "IssuerJwks": "",
> >     "TLSCertificate": "/etc/pelican/certificates/tls.crt",
> >     "TLSCACertificateFile": "/etc/pki/tls/cert.pem",
> >     "TLSCACertificateDirectory": "",
> >     "TLSCAKey": "/etc/pelican/certificates/tlsca.key",
> >     "ExternalAddress": "",
> >     "Hostname": "d92baeb5b3f7"
> >   },
> >   "Transport": {
> >     "ExpectContinueTimeout": 1000000000,
> >     "ResponseHeaderTimeout": 10000000000,
> >     "DialerTimeout": 10000000000,
> >     "DialerKeepAlive": 30000000000,
> >     "MaxIdleConns": 30,
> >     "IdleConnTimeout": 90000000000,
> >     "TLSHandshakeTimeout": 15000000000
> >   },
> >   "Xrootd": {
> >     "RobotsTxtFile": "/etc/pelican/robots.txt",
> >     "MacaroonsKeyFile": "/etc/pelican/macaroons-secret",
> >     "Authfile": "/etc/pelican/xrootd/authfile",
> >     "ManagerHost": "redirector.osgstorage.org",
> >     "SummaryMonitoringHost": "xrd-report.osgstorage.org",
> >     "DetailedMonitoringHost": "xrd-mon.osgstorage.org",
> >     "Port": 8443,
> >     "RunLocation": "/run/pelican/xrootd",
> >     "ScitokensConfig": "/etc/pelican/xrootd/scitokens.cfg",
> >     "Mount": "",
> >     "LocalMonitoringHost": "",
> >     "Sitename": "d92baeb5b3f7"
> >   }
> > }```
> > ```
> 
> Surprisingly, it seems that we never set the value of ConfigBase, even the default value in our code. Will ask other people why this is happening but it should be a bug not related to this PR. Thanks for catching it though!

ConfigBase is a special parameter, you'll notice it isn't created as a parameter in parameters.go, either. It's gotten using the function getConfigBase in config/config.go

haoming29 commented:
Based on Emma's input, I removed `ConfigBase` from REST API and added `ConfigDir` which is what actually being set in the code. @CannonLock you might have to change the field name in UI as well.

bbockelm reviewed (approved):
LGTM.
